datasources/firefox/firefox.go

```diff
@@ -0,0 +1,174 @@
+/*
+	Timelinize
+	Copyright (c) 2024 Sergio Rubio
+
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published
+	by the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+// Package firefox implements a data source that imports visited sites
+// from Firefox's places.sqlite database.
+package firefox
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/timelinize/timelinize/timeline"
+	"go.uber.org/zap"
+)
+
+// Data source name and ID.
+const (
+	DataSourceName = "Firefox"
+	DataSourceID   = "firefox"
+)
+
+func init() {
+	err := timeline.RegisterDataSource(timeline.DataSource{
+		Name:            DataSourceID,
+		Title:           DataSourceName,
+		Icon:            "firefox.svg",
+		Description:     "Firefox places.sqlite database importer",
+		NewFileImporter: func() timeline.FileImporter { return new(Firefox) },
+	})
+	if err != nil {
+		timeline.Log.Fatal("registering data source", zap.Error(err))
+	}
+}
+
+// Firefox interacts with the places.sqlite database to get visited sites.
+type Firefox struct{}
+
+// Recognize returns whether the input file is recognized.
+func (Firefox) Recognize(_ context.Context, dirEntry timeline.DirEntry, _ timeline.RecognizeParams) (timeline.Recognition, error) {
+	if filepath.Base(dirEntry.Name()) == "places.sqlite" {
+		return timeline.Recognition{Confidence: 1}, nil
```

Comment:
My only thought is, are you sure it's a Firefox DB? I mean, it probably is, but maybe not 100%. I'd set the confidence a little lower, unless you actually open the file and check that it has the needed schema.

Another option would be to look at the full path to see if it's in a Firefox folder (but, this isn't always a surefire solution either, since the file could have been moved/copied). Still, it could raise the confidence a bit if you don't want to open it.

Anyway, this can all be done later, and may never be and still work fine, I know there's plenty of data sources I've written where I am not super thorough with recognizing yet! :sweat_smile:

Anyway, just something to think about, but definitely not a blocker.

Reply:
> My only thought is, are you sure it's a Firefox DB? I mean, it probably is, but maybe not 100%. I'd set the confidence a little lower, unless you actually open the file and check that it has the needed schema.

I was thinking along those lines today. I could probably open the database and read available tables or the schema to confirm. Checking the path may be tricky given all the Firefox flavors and platforms.